File: backend/app/api/v1/arco.py
"""Endpoints públicos del Portal de Autoservicio ARCO (HU 3.2)."""
from fastapi import APIRouter, File, Form, UploadFile

from app.api.deps import DbSession
from app.schemas.arco import (
    ArcoAccessRequest,
    ArcoCancellationRequest,
    ArcoIdentityVerifyResponse,
    ArcoOppositionRequest,
    ArcoRectificationRequest,
    ArcoRequestRead,
)
from app.services.arco_service import ArcoService

router = APIRouter(prefix="/arco", tags=["arco"])


@router.post("/identity/verify", response_model=ArcoIdentityVerifyResponse, status_code=201)
def verify_identity(db: DbSession, dni: str = Form(..., min_length=8, max_length=8), identity_photo: UploadFile = File(...)):
    token = ArcoService(db).verify_identity(dni, identity_photo)
    return ArcoIdentityVerifyResponse(validation_token=token, dni=dni)


@router.post("/requests/access", response_model=ArcoRequestRead, status_code=201)
def request_access(payload: ArcoAccessRequest, db: DbSession):
    return ArcoService(db).request_access(payload.validation_token)


@router.post("/requests/rectification", response_model=ArcoRequestRead, status_code=201)
def request_rectification(payload: ArcoRectificationRequest, db: DbSession):
    return ArcoService(db).request_rectification(payload.validation_token, payload.fields_to_rectify, payload.reason)


@router.post("/requests/cancellation", response_model=ArcoRequestRead, status_code=201)
def request_cancellation(payload: ArcoCancellationRequest, db: DbSession):
    return ArcoService(db).request_cancellation(payload.validation_token, payload.reason, payload.confirm)


@router.post("/requests/opposition", response_model=ArcoRequestRead, status_code=201)
def request_opposition(payload: ArcoOppositionRequest, db: DbSession):
    return ArcoService(db).request_opposition(payload.validation_token, payload.reason)


@router.get("/requests/track/{dni}", response_model=list[ArcoRequestRead])
def track_requests(dni: str, db: DbSession):
    return ArcoService(db).track_requests(dni)

File: frontend/src/api/arco.ts
import axios from "axios";
import type { ArcoRequestRead } from "./types";

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL ?? "http://localhost:8000/api/v1";

// El Portal ARCO es público (HU 3.2): no requiere sesión ni clave API.
const arcoClient = axios.create({ baseURL: API_BASE_URL });

export async function verifyIdentity(dni: string, identityPhoto: File): Promise<{ validation_token: string; dni: string }> {
  const formData = new FormData();
  formData.append("dni", dni);
  formData.append("identity_photo", identityPhoto);
  const { data } = await arcoClient.post("/arco/identity/verify", formData, {
    headers: { "Content-Type": "multipart/form-data" },
  });
  return data;
}

export async function requestAccess(validation_token: string): Promise<ArcoRequestRead> {
  const { data } = await arcoClient.post("/arco/requests/access", { validation_token });
  return data;
}

export async function requestRectification(
  validation_token: string,
  fields_to_rectify: Record<string, string>,
  reason: string
): Promise<ArcoRequestRead> {
  const { data } = await arcoClient.post("/arco/requests/rectification", { validation_token, fields_to_rectify, reason });
  return data;
}

export async function requestCancellation(validation_token: string, reason: string, confirm: boolean): Promise<ArcoRequestRead> {
  const { data } = await arcoClient.post("/arco/requests/cancellation", { validation_token, reason, confirm });
  return data;
}

export async function requestOpposition(validation_token: string, reason: string): Promise<ArcoRequestRead> {
  const { data } = await arcoClient.post("/arco/requests/opposition", { validation_token, reason });
  return data;
}

export async function trackRequests(dni: string): Promise<ArcoRequestRead[]> {
  const { data } = await arcoClient.get(`/arco/requests/track/${dni}`);
  return data;
}
